[FEATURE] 동아리에서의 역할 기반 인가 시스템

jnujh opened the pull request:
## 🚀 작업 내용

이번 PR은 동아리 내 사용자의 역할(회장, 운영진, 일반 회원)에 따라 특정 API에 대한 접근을 제어하는 **역할 기반 인가(Role-Based Authorization) 시스템을 구현**하고, 그 안정성을 검증하기 위한 종합적인 테스트를 추가하는 것을 목표로 합니다.

**1. 역할 기반 인가 시스템 구현**
* **JWT 클레임 확장:**
    * `JwtProvider`를 수정하여, Access Token 발급 시 사용자의 모든 동아리 멤버십 정보(`{clubId: role}`)를 `memberships` 클레임에 포함하도록 했습니다.
    * 이를 통해 매 요청마다 DB를 조회하지 않고도 사용자의 역할 정보를 파악할 수 있습니다.
* **`PrincipalDetails` 확장:**
    * `memberships` 클레임 정보를 `PrincipalDetails` 객체에 저장하고, 이를 `CLUB_{clubId}_{ROLE}` 형식의 `GrantedAuthority`로 변환하도록 구현했습니다.
* **`@PreAuthorize`를 이용한 접근 제어:**
    * 각 컨트롤러의 API에 `@PreAuthorize` 어노테이션을 적용하여, 다음과 같은 두 가지 방식으로 접근 제어 규칙을 설정했습니다.
    * **기본 인가 규칙:** `hasAuthority('...')` 표현식을 사용하여, 특정 권한 문자열을 가진 사용자만 접근을 허용합니다. (적용 대상: `ClubController`, `ApplicationController`, `ClubApplyFormController`)
    * **동적 인가 규칙:** `@customSecurityService` 빈을 호출하여, DB 조회 등 복잡한 로직을 통해 동적으로 권한을 검사합니다. (적용 대상: `CommentController`)

**2. 인가 로직에 대한 통합 테스트 작성**
* **테스트 환경 구축 (`@SpringBootTest`):**
    * `@PreAuthorize`와 같은 메소드 시큐리티는 AOP를 통해 동작하므로, 웹 계층만 테스트하는 `@WebMvcTest` 환경에서는 동작하지 않는 문제가 있었습니다.
    * 이를 해결하기 위해, 실제 애플리케이션과 동일하게 모든 설정을 로드하는 `@SpringBootTest` 환경으로 전환하여 테스트의 신뢰도를 확보했습니다.
* **커스텀 테스트 어노테이션 (`@WithMockCustomUser`) 제작:**
    * `@customSecurityService`처럼 `PrincipalDetails`의 `memberships` 맵을 직접 사용하는 동적 인가 로직을 테스트하기 위해, 기존 `@WithMockUser`로는 한계가 있었습니다.
    * 이를 해결하기 위해, `memberships` 정보를 포함한 `PrincipalDetails` 객체를 테스트 컨텍스트에 직접 주입해주는 `@WithMockCustomUser` 커스텀 어노테이션과 팩토리를 직접 구현했습니다.
* **4가지 핵심 시나리오 검증:**
    * 모든 인가 API에 대해, 다음과 같은 4가지 핵심 시나리오를 모두 검증하는 테스트 케이스를 작성했습니다.
        1.  **성공 케이스:** 권한이 있는 사용자
        2.  **실패 케이스 (권한 부족):** 역할 등급이 낮은 사용자
        3.  **실패 케이스 (수평적 권한):** 다른 동아리의 권한을 가진 사용자
        4.  **실패 케이스 (미인증):** 로그인하지 않은 사용자

**3. 예외 처리 개선**
* `@PreAuthorize`에 의해 접근이 거부될 때 발생하는 `AuthorizationDeniedException`을 처리하는 핸들러를 `GlobalExceptionHandler`에 추가하여, 일관된 **403 Forbidden** 에러 응답이 반환되도록 수정했습니다.

---

## ⏱️ 소요 시간

5days

---

## 🤔 고민했던 내용

1.  **`@WebMvcTest` vs `@SpringBootTest`**
    * 초기에는 가벼운 `@WebMvcTest`로 인가 테스트를 시도했으나, `@PreAuthorize`가 동작하지 않는 문제를 겪었습니다. 원인 분석 결과, `@WebMvcTest`는 메소드 시큐리티를 위한 AOP 프록시를 기본적으로 생성하지 않는다는 것을 알게 되었습니다.
    * 이를 해결하기 위해, 실제 애플리케이션과 거의 동일한 환경을 제공하는 `@SpringBootTest`로 테스트 환경을 전환하여 문제의 근본 원인을 해결했습니다.

2.  **`@WithMockUser`의 한계와 `@WithMockCustomUser` 제작**
    * `@customSecurityService`와 같이 `PrincipalDetails`의 `memberships` 맵을 직접 사용하는 동적 인가 로직을 테스트할 때, `@WithMockUser`는 `authorities` 문자열만 제공할 뿐, `memberships` 정보를 포함한 `PrincipalDetails` 객체를 만들어주지 못하는 한계가 있었습니다.
    * 이 문제를 해결하기 위해, `WithSecurityContextFactory`를 사용하여 `memberships` 정보를 포함한 `PrincipalDetails` 객체를 테스트 컨텍스트에 직접 주입해주는 `@WithMockCustomUser` 커스텀 어노테이션을 직접 구현했습니다. 이 과정을 통해 Spring Security 테스트의 확장성에 대해 깊이 이해할 수 있었습니다.

3.  **API 명세와 실제 구현의 일치**
    * 개발 과정에서 프론트엔드 팀원의 피드백을 통해, API 명세의 에러 응답 형식이 실제 서버의 `ErrorResponseDto`와 다르다는 점을 발견했습니다.
    * 코드 리뷰 요청 전에, 모든 에러 응답 명세를 `{"error_code": "...", "message": "..."}` 형식으로 통일하여 문서와 코드의 일관성을 확보하는 작업을 선행했습니다.

---

## 💬 리뷰 중점사항

* `JwtProvider`에 `memberships` 클레임을 추가하고, `JwtAuthenticationFilter`에서 이를 파싱하여 `PrincipalDetails`를 생성하는 인증 흐름이 적절한지 검토 부탁드립니다.
* `@PreAuthorize`에 적용된 SpEL 표현식과 `@customSecurityService`의 동적 권한 검사 로직이 비즈니스 요구사항에 맞게 잘 구현되었는지 확인 부탁드립니다.
* `@SpringBootTest` 환경과 `@WithMockCustomUser`를 사용하여 인가 테스트를 구성한 전략이 효율적이고 올바른 접근 방식인지에 대한 피드백을 환영합니다.

---

## 🔗 관련 이슈

* Close #81

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

# 릴리스 노트

* **새로운 기능**
  * 동아리별 역할 기반 접근 제어 강화 및 JWT에 멤버십 정보 포함
  * 테스트용 커스텀 인증 어노테이션과 보안 유틸 추가

* **버그 수정 / 개선**
  * 인증·인가 흐름 정비로 권한 거부를 일관된 403 응답으로 처리
  * 댓글·신청 등에서 실제 인증 사용자 정보로 권한 검증 적용
  * 공개/보호 엔드포인트 범위 명확화

* **테스트**
  * 권한 관련 통합·단위 테스트 대폭 추가/갱신

* **기타**
  * 기본 샘플 데이터 전면 교체 및 알림 엔티티 제거
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jnujh committed:
feat: 역할 정보 조회를 위한 ClubMembershipInfo DTO 추가 (#81)

역할 기반 인가 구현의 일환으로, 데이터베이스 조회 시 동아리 ID(clubId)와 역할(role) 정보만을 선택적으로 담기 위한 DTO를 추가합니다.
이 DTO는 향후 Repository에서 JPQL 생성자 표현식과 함께 사용되어,
엔티티 전체가 아닌 필요한 데이터만 조회(Projection)함으로써 성능을 최적화하는 데 사용됩니다.

jnujh committed:
feat: 역할 정보 조회를 위한 ClubMembershipInfo DTO 추가 (#81)

역할 기반 인가 구현의 일환으로, 데이터베이스 조회 시 동아리 ID(clubId)와 역할(role) 정보만을 선택적으로 담기 위한 DTO를 추가합니다.
이 DTO는 향후 Repository에서 JPQL 생성자 표현식과 함께 사용되어,
엔티티 전체가 아닌 필요한 데이터만 조회(Projection)함으로써 성능을 최적화하는 데 사용됩니다.

jnujh committed:
feat: 역할 정보 조회를 위한 ClubMembershipInfo DTO 추가 (#81)

역할 기반 인가 구현의 일환으로, 데이터베이스 조회 시 동아리 ID(clubId)와 역할(role) 정보만을 선택적으로 담기 위한 DTO를 추가합니다.
이 DTO는 향후 Repository에서 JPQL 생성자 표현식과 함께 사용되어,
엔티티 전체가 아닌 필요한 데이터만 조회(Projection)함으로써 성능을 최적화하는 데 사용됩니다.

jnujh committed:
feat: 사용자의 모든 역할 정보를 DTO로 조회하는 메소드 추가 (#81)

특정 사용자가 속한 모든 동아리의 ID와 역할 정보를 ClubMembershipInfo DTO 리스트로 직접 반환하는 기능을 추가합니다.
- findClubMembershipsByUser 메소드를 새로 정의했습니다.
- JPQL의 select new 생성자 표현식을 사용하여, 조회 결과를 엔티티를 거치지 않고 즉시 DTO로 매핑합니다.
- 이 방식은 필요한 컬럼만 조회하여 네트워크 트래픽을 줄이고, ORM의 엔티티 관리 부하를 피할 수 있습니다.

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/domain/clubMember/repository/ClubMemberRepository.java

jnujh committed:
feat: 사용자의 모든 역할 정보를 DTO로 조회하는 메소드 추가 (#81)

특정 사용자가 속한 모든 동아리의 ID와 역할 정보를 ClubMembershipInfo DTO 리스트로 직접 반환하는 기능을 추가합니다.
- findClubMembershipsByUser 메소드를 새로 정의했습니다.
- JPQL의 select new 생성자 표현식을 사용하여, 조회 결과를 엔티티를 거치지 않고 즉시 DTO로 매핑합니다.
- 이 방식은 필요한 컬럼만 조회하여 네트워크 트래픽을 줄이고, ORM의 엔티티 관리 부하를 피할 수 있습니다.

jnujh committed:
feat: 사용자의 모든 역할 정보를 DTO로 조회하는 메소드 추가 (#81)

특정 사용자가 속한 모든 동아리의 ID와 역할 정보를 ClubMembershipInfo DTO 리스트로 직접 반환하는 기능을 추가합니다.
- findClubMembershipsByUser 메소드를 새로 정의했습니다.
- JPQL의 select new 생성자 표현식을 사용하여, 조회 결과를 엔티티를 거치지 않고 즉시 DTO로 매핑합니다.
- 이 방식은 필요한 컬럼만 조회하여 네트워크 트래픽을 줄이고, ORM의 엔티티 관리 부하를 피할 수 있습니다.

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/domain/clubMember/repository/ClubMemberRepository.java

jnujh committed:
feat: 사용자의 모든 역할 정보를 DTO로 조회하는 메소드 추가 (#81)

특정 사용자가 속한 모든 동아리의 ID와 역할 정보를 ClubMembershipInfo DTO 리스트로 직접 반환하는 기능을 추가합니다.
- findClubMembershipsByUser 메소드를 새로 정의했습니다.
- JPQL의 select new 생성자 표현식을 사용하여, 조회 결과를 엔티티를 거치지 않고 즉시 DTO로 매핑합니다.
- 이 방식은 필요한 컬럼만 조회하여 네트워크 트래픽을 줄이고, ORM의 엔티티 관리 부하를 피할 수 있습니다.

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/domain/clubMember/repository/ClubMemberRepository.java

jnujh committed:
feat: JWT 액세스 토큰에 동아리 멤버십(memberships) 클레임 추가 (#81)

사용자 인증 시 생성되는 JWT 액세스 토큰에 동아리별 역할 정보를 포함하도록 수정합니다.
- JwtProvider.createAccessToken메소드가 새로 추가된findClubMembershipsByUser를 호출하도록 변경했습니다.
- 조회된 List<ClubMembershipInfo>를 {clubId: roleName} 형태의 Map으로 변환하여,
"memberships"라는 이름의 클레임으로 토큰에 추가합니다.
- 이를 통해 토큰 자체에 인가(Authorization)에 필요한 역할 정보가 담기게 되어, 향후 API 접근 제어의 기반이 마련되었습니다.

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/global/security/JwtProvider.java

jnujh committed:
feat: JWT 액세스 토큰에 동아리 멤버십(memberships) 클레임 추가 (#81)

사용자 인증 시 생성되는 JWT 액세스 토큰에 동아리별 역할 정보를 포함하도록 수정합니다.
- JwtProvider.createAccessToken메소드가 새로 추가된findClubMembershipsByUser를 호출하도록 변경했습니다.
- 조회된 List<ClubMembershipInfo>를 {clubId: roleName} 형태의 Map으로 변환하여,
"memberships"라는 이름의 클레임으로 토큰에 추가합니다.
- 이를 통해 토큰 자체에 인가(Authorization)에 필요한 역할 정보가 담기게 되어, 향후 API 접근 제어의 기반이 마련되었습니다.

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/global/security/JwtProvider.java

jnujh committed:
feat: JWT 액세스 토큰에 동아리 멤버십(memberships) 클레임 추가 (#81)

사용자 인증 시 생성되는 JWT 액세스 토큰에 동아리별 역할 정보를 포함하도록 수정합니다.
- JwtProvider.createAccessToken메소드가 새로 추가된findClubMembershipsByUser를 호출하도록 변경했습니다.
- 조회된 List<ClubMembershipInfo>를 {clubId: roleName} 형태의 Map으로 변환하여,
"memberships"라는 이름의 클레임으로 토큰에 추가합니다.
- 이를 통해 토큰 자체에 인가(Authorization)에 필요한 역할 정보가 담기게 되어, 향후 API 접근 제어의 기반이 마련되었습니다.

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/global/security/JwtProvider.java

jnujh committed:
feat: PrincipalDetails에 동아리 역할 정보(memberships) 필드 추가 (#81)

Spring Security가 관리하는 사용자 정보 객체(Principal)에 동아리별 역할 정보를 저장하기 위한 필드를 추가합니다.
- final타입의 Map<String, String> memberships 필드를 추가하여,
JWT에서 추출한 역할 정보를 담을 수 있도록 구조를 확장했습니다.
- 이를 통해, 인증된 사용자의 역할 정보가 애플리케이션 전반의 SecurityContext 내에서 접근 가능하게 됩니다.

jnujh committed:
feat: PrincipalDetails에 동아리 역할 정보(memberships) 필드 추가 (#81)

Spring Security가 관리하는 사용자 정보 객체(Principal)에 동아리별 역할 정보를 저장하기 위한 필드를 추가합니다.
- final타입의 Map<String, String> memberships 필드를 추가하여,
JWT에서 추출한 역할 정보를 담을 수 있도록 구조를 확장했습니다.
- 이를 통해, 인증된 사용자의 역할 정보가 애플리케이션 전반의 SecurityContext 내에서 접근 가능하게 됩니다.

jnujh committed:
feat: PrincipalDetails에 동아리 역할 정보(memberships) 필드 추가 (#81)

Spring Security가 관리하는 사용자 정보 객체(Principal)에 동아리별 역할 정보를 저장하기 위한 필드를 추가합니다.
- final타입의 Map<String, String> memberships 필드를 추가하여,
JWT에서 추출한 역할 정보를 담을 수 있도록 구조를 확장했습니다.
- 이를 통해, 인증된 사용자의 역할 정보가 애플리케이션 전반의 SecurityContext 내에서 접근 가능하게 됩니다.

jnujh committed:
refactor: PrincipalDetailsService가 역할 정보를 주입받도록 수정 (#81)

JWT 토큰의 역할 정보를 PrincipalDetails 객체에 주입하기 위해 서비스 로직을 개선합니다.
- loadUserByUsername 메소드를 오버로딩하여,memberships맵을 인자로 받는 새로운 메소드를 추가했습니다.
이 메소드가PrincipalDetails 생성 시 역할 정보를 함께 전달하는 핵심 역할을 합니다.
- UserDetailsService 인터페이스 계약을 위한 기존 메소드는 @Deprecated 처리하고,
Javadoc을 통해 새로운 메소드 사용을 권장하도록 명시했습니다.
- 이를 통해 프레임워크와의 호환성을 유지하면서, 의도치 않은 사용을 방지하고 코드의 역할을 명확히 분리했습니다.

jnujh committed:
refactor: PrincipalDetailsService가 역할 정보를 주입받도록 수정 (#81)

JWT 토큰의 역할 정보를 PrincipalDetails 객체에 주입하기 위해 서비스 로직을 개선합니다.
- loadUserByUsername 메소드를 오버로딩하여,memberships맵을 인자로 받는 새로운 메소드를 추가했습니다.
이 메소드가PrincipalDetails 생성 시 역할 정보를 함께 전달하는 핵심 역할을 합니다.
- UserDetailsService 인터페이스 계약을 위한 기존 메소드는 @Deprecated 처리하고,
Javadoc을 통해 새로운 메소드 사용을 권장하도록 명시했습니다.
- 이를 통해 프레임워크와의 호환성을 유지하면서, 의도치 않은 사용을 방지하고 코드의 역할을 명확히 분리했습니다.

jnujh committed:
refactor: PrincipalDetailsService가 역할 정보를 주입받도록 수정 (#81)

JWT 토큰의 역할 정보를 PrincipalDetails 객체에 주입하기 위해 서비스 로직을 개선합니다.
- loadUserByUsername 메소드를 오버로딩하여,memberships맵을 인자로 받는 새로운 메소드를 추가했습니다.
이 메소드가PrincipalDetails 생성 시 역할 정보를 함께 전달하는 핵심 역할을 합니다.
- UserDetailsService 인터페이스 계약을 위한 기존 메소드는 @Deprecated 처리하고,
Javadoc을 통해 새로운 메소드 사용을 권장하도록 명시했습니다.
- 이를 통해 프레임워크와의 호환성을 유지하면서, 의도치 않은 사용을 방지하고 코드의 역할을 명확히 분리했습니다.

jnujh committed:
feat: JwtAuthenticationFilter가 토큰의 역할 정보를 파싱하도록 수정 (#81)

API 요청의 JWT 토큰에서 역할 정보를 추출하여, 인증된 사용자 정보에 포함시키는 기능을 추가합니다.
- 토큰 검증 후, 페이로드에서 "memberships" 클레임을 Map 형태로 추출하는 로직을 추가했습니다.
- PrincipalDetailsService의 새로운 loadUserByUsername(username, memberships)메소드를 호출하도록 변경하여,
토큰의 역할 정보가 PrincipalDetails 객체에 최종적으로 담기도록 연결했습니다.
- 이로써, 토큰에 담긴 역할 정보가 SecurityContext에 저장되기까지의 모든 과정이 완료되었습니다.

jnujh committed:
feat: JwtAuthenticationFilter가 토큰의 역할 정보를 파싱하도록 수정 (#81)

API 요청의 JWT 토큰에서 역할 정보를 추출하여, 인증된 사용자 정보에 포함시키는 기능을 추가합니다.
- 토큰 검증 후, 페이로드에서 "memberships" 클레임을 Map 형태로 추출하는 로직을 추가했습니다.
- PrincipalDetailsService의 새로운 loadUserByUsername(username, memberships)메소드를 호출하도록 변경하여,
토큰의 역할 정보가 PrincipalDetails 객체에 최종적으로 담기도록 연결했습니다.
- 이로써, 토큰에 담긴 역할 정보가 SecurityContext에 저장되기까지의 모든 과정이 완료되었습니다.

jnujh committed:
feat: JwtAuthenticationFilter가 토큰의 역할 정보를 파싱하도록 수정 (#81)

API 요청의 JWT 토큰에서 역할 정보를 추출하여, 인증된 사용자 정보에 포함시키는 기능을 추가합니다.
- 토큰 검증 후, 페이로드에서 "memberships" 클레임을 Map 형태로 추출하는 로직을 추가했습니다.
- PrincipalDetailsService의 새로운 loadUserByUsername(username, memberships)메소드를 호출하도록 변경하여,
토큰의 역할 정보가 PrincipalDetails 객체에 최종적으로 담기도록 연결했습니다.
- 이로써, 토큰에 담긴 역할 정보가 SecurityContext에 저장되기까지의 모든 과정이 완료되었습니다.

jnujh committed:
feat: JwtAuthenticationFilter가 토큰의 역할 정보를 파싱하도록 수정 (#81)

API 요청의 JWT 토큰에서 역할 정보를 추출하여, 인증된 사용자 정보에 포함시키는 기능을 추가합니다.
- 토큰 검증 후, 페이로드에서 "memberships" 클레임을 Map 형태로 추출하는 로직을 추가했습니다.
- PrincipalDetailsService의 새로운 loadUserByUsername(username, memberships)메소드를 호출하도록 변경하여,
토큰의 역할 정보가 PrincipalDetails 객체에 최종적으로 담기도록 연결했습니다.
- 이로써, 토큰에 담긴 역할 정보가 SecurityContext에 저장되기까지의 모든 과정이 완료되었습니다.

jnujh committed:
feat: PrincipalDetails에 역할 정보 및 권한(Authorities) 추가

역할 기반 인가를 위해, Spring Security의 사용자 정보 객체인 PrincipalDetails를 확장합니다.
- JWT에서 추출한 동아리별 역할 정보(memberships)를 저장하기 위한 Map 필드를 추가했습니다.
- getAuthorities()메소드를 구현하여,memberships맵을 'CLUB_{clubId}_{ROLE}'형식의
GrantedAuthority목록으로 변환합니다. 이는@PreAuthorize가 권한을 검사하는 기반이 됩니다.

jnujh committed:
feat: PrincipalDetails에 역할 정보 및 권한(Authorities) 추가

역할 기반 인가를 위해, Spring Security의 사용자 정보 객체인 PrincipalDetails를 확장합니다.
- JWT에서 추출한 동아리별 역할 정보(memberships)를 저장하기 위한 Map 필드를 추가했습니다.
- getAuthorities()메소드를 구현하여,memberships맵을 'CLUB_{clubId}_{ROLE}'형식의
GrantedAuthority목록으로 변환합니다. 이는@PreAuthorize가 권한을 검사하는 기반이 됩니다.

jnujh committed:
feat: PrincipalDetails에 역할 정보 및 권한(Authorities) 추가

역할 기반 인가를 위해, Spring Security의 사용자 정보 객체인 PrincipalDetails를 확장합니다.
- JWT에서 추출한 동아리별 역할 정보(memberships)를 저장하기 위한 Map 필드를 추가했습니다.
- getAuthorities()메소드를 구현하여,memberships맵을 'CLUB_{clubId}_{ROLE}'형식의
GrantedAuthority목록으로 변환합니다. 이는@PreAuthorize가 권한을 검사하는 기반이 됩니다.

jnujh committed:
feat: applicationId로 clubId를 조회하는 최적화 쿼리 추가

CustomSecurityService에서 지원서에 대한 권한을 검사할 때 발생하는 N+1 문제를 방지하고
성능을 최적화하기 위해 새로운 조회 메소드를 추가합니다.
- findClubIdByApplicationId 메소드를 추가했습니다.
- JPQL의 프로젝션(Projection)을 사용하여 엔티티 전체를 로딩하지 않고, 필요한 clubId 값만 직접 조회하여 DB 부하를 최소화합니다.

jnujh committed:
feat: applicationId로 clubId를 조회하는 최적화 쿼리 추가

CustomSecurityService에서 지원서에 대한 권한을 검사할 때 발생하는 N+1 문제를 방지하고
성능을 최적화하기 위해 새로운 조회 메소드를 추가합니다.
- findClubIdByApplicationId 메소드를 추가했습니다.
- JPQL의 프로젝션(Projection)을 사용하여 엔티티 전체를 로딩하지 않고, 필요한 clubId 값만 직접 조회하여 DB 부하를 최소화합니다.

jnujh committed:
feat: applicationId로 clubId를 조회하는 최적화 쿼리 추가

CustomSecurityService에서 지원서에 대한 권한을 검사할 때 발생하는 N+1 문제를 방지하고
성능을 최적화하기 위해 새로운 조회 메소드를 추가합니다.
- findClubIdByApplicationId 메소드를 추가했습니다.
- JPQL의 프로젝션(Projection)을 사용하여 엔티티 전체를 로딩하지 않고, 필요한 clubId 값만 직접 조회하여 DB 부하를 최소화합니다.

jnujh committed:
feat: 동적 권한 검사를 위한 CustomSecurityService 추가

@PreAuthorize 어노테이션 내에서 직접 처리하기 어려운 복잡한 인가 로직을 위임하기 위한 서비스를 추가합니다.
- API 시그니처에 clubId가 없는 경우, applicationId등으로 DB를 조회하여 권한을 검사하는 메소드를 포함합니다.
(e.g.,isClubAdminOrExecutiveForApplication)
- 성능 최적화를 위해, 새로 추가된 ApplicationRepository.findClubIdByApplicationId 메소드를 사용합니다.

jnujh committed:
feat: 동적 권한 검사를 위한 CustomSecurityService 추가

@PreAuthorize 어노테이션 내에서 직접 처리하기 어려운 복잡한 인가 로직을 위임하기 위한 서비스를 추가합니다.
- API 시그니처에 clubId가 없는 경우, applicationId등으로 DB를 조회하여 권한을 검사하는 메소드를 포함합니다.
(e.g.,isClubAdminOrExecutiveForApplication)
- 성능 최적화를 위해, 새로 추가된 ApplicationRepository.findClubIdByApplicationId 메소드를 사용합니다.

jnujh committed:
feat: 동적 권한 검사를 위한 CustomSecurityService 추가

@PreAuthorize 어노테이션 내에서 직접 처리하기 어려운 복잡한 인가 로직을 위임하기 위한 서비스를 추가합니다.
- API 시그니처에 clubId가 없는 경우, applicationId등으로 DB를 조회하여 권한을 검사하는 메소드를 포함합니다.
(e.g.,isClubAdminOrExecutiveForApplication)
- 성능 최적화를 위해, 새로 추가된 ApplicationRepository.findClubIdByApplicationId 메소드를 사용합니다.

jnujh committed:
feat: 동아리 관리 API에 역할 기반 인가 적용

updateClub, getClubDashboard 등 동아리 운영진을 위한 API에 CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할만
접근 가능하도록 @PreAuthorize를 사용한 인가 규칙을 추가합니다.

jnujh committed:
feat: 동아리 관리 API에 역할 기반 인가 적용

updateClub, getClubDashboard 등 동아리 운영진을 위한 API에 CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할만
접근 가능하도록 @PreAuthorize를 사용한 인가 규칙을 추가합니다.

jnujh committed:
feat: 동아리 관리 API에 역할 기반 인가 적용

updateClub, getClubDashboard 등 동아리 운영진을 위한 API에 CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할만
접근 가능하도록 @PreAuthorize를 사용한 인가 규칙을 추가합니다.

jnujh committed:
refactor: updateApplicationStatus URL 변경 및 인가 적용

updateApplicationStatus API의 URL에 clubId를 포함하도록 변경하여,
CustomSecurityService를 통한 추가 DB 조회 없이 @PreAuthorize에서 직접 권한을 검사할 수 있도록 개선합니다.
`- **URL 변경**: /api/applications/{appId}->/api/clubs/{clubId}/applications/{appId}/status``
- 이를 통해 성능을 향상시키고, 리소스의 계층 구조를 명확히 하는 RESTful 디자인을 따릅니다.
- getApplicationDetail과 변경된 updateApplicationStatus 두 메소드에
CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할만 접근 가능하도록 인가 규칙을 추가했습니다.

jnujh committed:
refactor: updateApplicationStatus URL 변경 및 인가 적용

updateApplicationStatus API의 URL에 clubId를 포함하도록 변경하여,
CustomSecurityService를 통한 추가 DB 조회 없이 @PreAuthorize에서 직접 권한을 검사할 수 있도록 개선합니다.
`- **URL 변경**: /api/applications/{appId}->/api/clubs/{clubId}/applications/{appId}/status``
- 이를 통해 성능을 향상시키고, 리소스의 계층 구조를 명확히 하는 RESTful 디자인을 따릅니다.
- getApplicationDetail과 변경된 updateApplicationStatus 두 메소드에
CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할만 접근 가능하도록 인가 규칙을 추가했습니다.

jnujh committed:
refactor: updateApplicationStatus URL 변경 및 인가 적용

updateApplicationStatus API의 URL에 clubId를 포함하도록 변경하여,
CustomSecurityService를 통한 추가 DB 조회 없이 @PreAuthorize에서 직접 권한을 검사할 수 있도록 개선합니다.
`- **URL 변경**: /api/applications/{appId}->/api/clubs/{clubId}/applications/{appId}/status``
- 이를 통해 성능을 향상시키고, 리소스의 계층 구조를 명확히 하는 RESTful 디자인을 따릅니다.
- getApplicationDetail과 변경된 updateApplicationStatus 두 메소드에
CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할만 접근 가능하도록 인가 규칙을 추가했습니다.

jnujh committed:
feat: 댓글 API에 인가 적용 및 사용자 ID 로직 수정

댓글 관련 API에 역할 기반 인가를 적용하고, 하드코딩된 사용자 ID 문제를 해결합니다.
- 모든 댓글 API에 대해, CustomSecurityService를 호출하여 해당 동아리의 운영진인지 검사하는 @PreAuthorize 인가 규칙을 추가합니다.
- 기존에 하드코딩되어 있던 사용자 ID(1L)를, @AuthenticationPrincipal을 사용하여 실제 인증된 사용자의 ID를 가져오도록 수정합니다.

jnujh committed:
feat: 댓글 API에 인가 적용 및 사용자 ID 로직 수정

댓글 관련 API에 역할 기반 인가를 적용하고, 하드코딩된 사용자 ID 문제를 해결합니다.
- 모든 댓글 API에 대해, CustomSecurityService를 호출하여 해당 동아리의 운영진인지 검사하는 @PreAuthorize 인가 규칙을 추가합니다.
- 기존에 하드코딩되어 있던 사용자 ID(1L)를, @AuthenticationPrincipal을 사용하여 실제 인증된 사용자의 ID를 가져오도록 수정합니다.

jnujh committed:
feat: 댓글 API에 인가 적용 및 사용자 ID 로직 수정

댓글 관련 API에 역할 기반 인가를 적용하고, 하드코딩된 사용자 ID 문제를 해결합니다.
- 모든 댓글 API에 대해, CustomSecurityService를 호출하여 해당 동아리의 운영진인지 검사하는 @PreAuthorize 인가 규칙을 추가합니다.
- 기존에 하드코딩되어 있던 사용자 ID(1L)를, @AuthenticationPrincipal을 사용하여 실제 인증된 사용자의 ID를 가져오도록 수정합니다.

jnujh committed:
feat: 지원서 양식 API에 역할 기반 인가 적용

CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할만 지원서 양식을 조회, 생성, 수정할 수 있도록 @PreAuthorize를 사용한 인가 규칙을 추가합니다.
- #clubId 경로 변수를 직접 참조하여 DB 조회 없이 효율적으로 권한을 검사합니다.
- 기존의 권한 관련 TODO 주석을 제거합니다.

jnujh committed:
feat: 지원서 양식 API에 역할 기반 인가 적용

CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할만 지원서 양식을 조회, 생성, 수정할 수 있도록 @PreAuthorize를 사용한 인가 규칙을 추가합니다.
- #clubId 경로 변수를 직접 참조하여 DB 조회 없이 효율적으로 권한을 검사합니다.
- 기존의 권한 관련 TODO 주석을 제거합니다.

jnujh committed:
feat: 지원서 양식 API에 역할 기반 인가 적용

CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할만 지원서 양식을 조회, 생성, 수정할 수 있도록 @PreAuthorize를 사용한 인가 규칙을 추가합니다.
- #clubId 경로 변수를 직접 참조하여 DB 조회 없이 효율적으로 권한을 검사합니다.
- 기존의 권한 관련 TODO 주석을 제거합니다.

jnujh committed:
test: API 엔드포인트 불일치로 인한 테스트 실패 수정

`ApplicationControllerTest`의 `updateApplicationStatus` 관련 테스트 실행 시, `NoResourceFoundException`이 발생했습니다.
이는 테스트 코드에서 호출하는 API 주소(`PATCH /api/applications/{applicationId}`)가
실제 컨트롤러에 정의된 주소(`PATCH /api/clubs/{clubId}/applications/{applicationId}/status`)와 일치하지 않았기 때문입니다.

`mockMvc.perform()` 호출 시, 실제 컨트롤러의 엔드포인트와 일치하도록 URL 경로를 수정하여 테스트가 정상적으로 동작하도록 수정했습니다.

jnujh committed:
test: API 엔드포인트 불일치로 인한 테스트 실패 수정

`ApplicationControllerTest`의 `updateApplicationStatus` 관련 테스트 실행 시, `NoResourceFoundException`이 발생했습니다.
이는 테스트 코드에서 호출하는 API 주소(`PATCH /api/applications/{applicationId}`)가
실제 컨트롤러에 정의된 주소(`PATCH /api/clubs/{clubId}/applications/{applicationId}/status`)와 일치하지 않았기 때문입니다.

`mockMvc.perform()` 호출 시, 실제 컨트롤러의 엔드포인트와 일치하도록 URL 경로를 수정하여 테스트가 정상적으로 동작하도록 수정했습니다.

jnujh committed:
test: API 엔드포인트 불일치로 인한 테스트 실패 수정

`ApplicationControllerTest`의 `updateApplicationStatus` 관련 테스트 실행 시, `NoResourceFoundException`이 발생했습니다.
이는 테스트 코드에서 호출하는 API 주소(`PATCH /api/applications/{applicationId}`)가
실제 컨트롤러에 정의된 주소(`PATCH /api/clubs/{clubId}/applications/{applicationId}/status`)와 일치하지 않았기 때문입니다.

`mockMvc.perform()` 호출 시, 실제 컨트롤러의 엔드포인트와 일치하도록 URL 경로를 수정하여 테스트가 정상적으로 동작하도록 수정했습니다.

jnujh committed:
feat: API 엔드포인트에 따른 접근 제어 규칙 구체화 (#81)

기존 SecurityConfig는 개발 편의를 위해 모든 API 요청(`anyRequest().permitAll()`)을 허용하고 있었습니다.
역할 기반 인가 테스트 및 실제 운영 환경을 준비하기 위해, 보안의 "최소 권한의 원칙"에 따라 API 접근 제어 규칙을 구체화할 필요가 있었습니다.

1.  기본 규칙을 `anyRequest().authenticated()`로 변경하여, 명시적으로 허용되지 않은 모든 요청은 인증을 요구하도록 강화했습니다.
2.  로그인하지 않은 사용자도 접근할 수 있어야 하는 공개 API들을 `requestMatchers`와 `HttpMethod`를 사용하여 명시적으로 허용(`permitAll`)했습니다.

    - `GET /api/clubs`, `GET /api/clubs/*` (동아리 목록/상세 조회)
    - `GET /api/clubs/*/apply` (지원서 양식 조회)
    - `POST /api/clubs/*/apply-submit` (지원서 제출)
    - `/api/auth/**`, Swagger 관련 경로 등

jnujh committed:
feat: API 엔드포인트에 따른 접근 제어 규칙 구체화 (#81)

기존 SecurityConfig는 개발 편의를 위해 모든 API 요청(`anyRequest().permitAll()`)을 허용하고 있었습니다.
역할 기반 인가 테스트 및 실제 운영 환경을 준비하기 위해, 보안의 "최소 권한의 원칙"에 따라 API 접근 제어 규칙을 구체화할 필요가 있었습니다.

1.  기본 규칙을 `anyRequest().authenticated()`로 변경하여, 명시적으로 허용되지 않은 모든 요청은 인증을 요구하도록 강화했습니다.
2.  로그인하지 않은 사용자도 접근할 수 있어야 하는 공개 API들을 `requestMatchers`와 `HttpMethod`를 사용하여 명시적으로 허용(`permitAll`)했습니다.

    - `GET /api/clubs`, `GET /api/clubs/*` (동아리 목록/상세 조회)
    - `GET /api/clubs/*/apply` (지원서 양식 조회)
    - `POST /api/clubs/*/apply-submit` (지원서 제출)
    - `/api/auth/**`, Swagger 관련 경로 등

jnujh committed:
feat: API 엔드포인트에 따른 접근 제어 규칙 구체화 (#81)

기존 SecurityConfig는 개발 편의를 위해 모든 API 요청(`anyRequest().permitAll()`)을 허용하고 있었습니다.
역할 기반 인가 테스트 및 실제 운영 환경을 준비하기 위해, 보안의 "최소 권한의 원칙"에 따라 API 접근 제어 규칙을 구체화할 필요가 있었습니다.

1.  기본 규칙을 `anyRequest().authenticated()`로 변경하여, 명시적으로 허용되지 않은 모든 요청은 인증을 요구하도록 강화했습니다.
2.  로그인하지 않은 사용자도 접근할 수 있어야 하는 공개 API들을 `requestMatchers`와 `HttpMethod`를 사용하여 명시적으로 허용(`permitAll`)했습니다.

    - `GET /api/clubs`, `GET /api/clubs/*` (동아리 목록/상세 조회)
    - `GET /api/clubs/*/apply` (지원서 양식 조회)
    - `POST /api/clubs/*/apply-submit` (지원서 제출)
    - `/api/auth/**`, Swagger 관련 경로 등

jnujh committed:
feat: API 엔드포인트에 따른 접근 제어 규칙 구체화 (#81)

기존 SecurityConfig는 개발 편의를 위해 모든 API 요청(`anyRequest().permitAll()`)을 허용하고 있었습니다.
역할 기반 인가 테스트 및 실제 운영 환경을 준비하기 위해, 보안의 "최소 권한의 원칙"에 따라 API 접근 제어 규칙을 구체화할 필요가 있었습니다.

1.  기본 규칙을 `anyRequest().authenticated()`로 변경하여, 명시적으로 허용되지 않은 모든 요청은 인증을 요구하도록 강화했습니다.
2.  로그인하지 않은 사용자도 접근할 수 있어야 하는 공개 API들을 `requestMatchers`와 `HttpMethod`를 사용하여 명시적으로 허용(`permitAll`)했습니다.

    - `GET /api/clubs`, `GET /api/clubs/*` (동아리 목록/상세 조회)
    - `GET /api/clubs/*/apply` (지원서 양식 조회)
    - `POST /api/clubs/*/apply-submit` (지원서 제출)
    - `/api/auth/**`, Swagger 관련 경로 등

jnujh committed:
feat: 메소드 레벨 보안 기능 활성화 (#81)

[변경 배경]
@SpringBootTest 환경에서 @PreAuthorize 어노테이션이 동작하지 않는 문제가 있었습니다.
이는 메소드 시큐리티 기능이 활성화되지 않았기 때문입니다.

[변경 내용]
SecurityConfig에 @EnableMethodSecurity 어노테이션을 추가하여,
@PreAuthorize, @PostAuthorize 등과 같은 메소드 레벨의 보안 어노테이션이 AOP를 통해 정상적으로 동작하도록 설정했습니다.

[기대 효과]
이 변경으로 인해 컨트롤러 및 서비스 계층에서 메소드 단위의 동적 권한 검사가 가능해집니다.

jnujh committed:
feat: 메소드 레벨 보안 기능 활성화 (#81)

[변경 배경]
@SpringBootTest 환경에서 @PreAuthorize 어노테이션이 동작하지 않는 문제가 있었습니다.
이는 메소드 시큐리티 기능이 활성화되지 않았기 때문입니다.

[변경 내용]
SecurityConfig에 @EnableMethodSecurity 어노테이션을 추가하여,
@PreAuthorize, @PostAuthorize 등과 같은 메소드 레벨의 보안 어노테이션이 AOP를 통해 정상적으로 동작하도록 설정했습니다.

[기대 효과]
이 변경으로 인해 컨트롤러 및 서비스 계층에서 메소드 단위의 동적 권한 검사가 가능해집니다.

jnujh committed:
feat: 메소드 레벨 보안 기능 활성화 (#81)

[변경 배경]
@SpringBootTest 환경에서 @PreAuthorize 어노테이션이 동작하지 않는 문제가 있었습니다.
이는 메소드 시큐리티 기능이 활성화되지 않았기 때문입니다.

[변경 내용]
SecurityConfig에 @EnableMethodSecurity 어노테이션을 추가하여,
@PreAuthorize, @PostAuthorize 등과 같은 메소드 레벨의 보안 어노테이션이 AOP를 통해 정상적으로 동작하도록 설정했습니다.

[기대 효과]
이 변경으로 인해 컨트롤러 및 서비스 계층에서 메소드 단위의 동적 권한 검사가 가능해집니다.

jnujh committed:
feat: 메소드 레벨 보안 기능 활성화 (#81)

[변경 배경]
@SpringBootTest 환경에서 @PreAuthorize 어노테이션이 동작하지 않는 문제가 있었습니다.
이는 메소드 시큐리티 기능이 활성화되지 않았기 때문입니다.

[변경 내용]
SecurityConfig에 @EnableMethodSecurity 어노테이션을 추가하여,
@PreAuthorize, @PostAuthorize 등과 같은 메소드 레벨의 보안 어노테이션이 AOP를 통해 정상적으로 동작하도록 설정했습니다.

[기대 효과]
이 변경으로 인해 컨트롤러 및 서비스 계층에서 메소드 단위의 동적 권한 검사가 가능해집니다.

jnujh committed:
fix: 메소드 시큐리티 인가 예외 처리 핸들러 추가 (#81)

[문제 원인]
@PreAuthorize 어노테이션에 의해 접근이 거부될 때, Spring Security는 `AuthorizationDeniedException`을 발생시킵니다.
이 예외를 처리하는 핸들러가 없어, 최종적으로 `500 Internal Server Error`가 응답되는 문제가 있었습니다.

[해결 방안]
GlobalExceptionHandler에 `AuthorizationDeniedException`을 처리하는 @ExceptionHandler 메소드를 추가했습니다.
이 핸들러는 해당 예외를 우리가 정의한 `ForbiddenAccessException`으로 변환하여, 일관된 403 Forbidden 에러 응답(JSON)이 반환되도록 합니다.

jnujh committed:
fix: 메소드 시큐리티 인가 예외 처리 핸들러 추가 (#81)

[문제 원인]
@PreAuthorize 어노테이션에 의해 접근이 거부될 때, Spring Security는 `AuthorizationDeniedException`을 발생시킵니다.
이 예외를 처리하는 핸들러가 없어, 최종적으로 `500 Internal Server Error`가 응답되는 문제가 있었습니다.

[해결 방안]
GlobalExceptionHandler에 `AuthorizationDeniedException`을 처리하는 @ExceptionHandler 메소드를 추가했습니다.
이 핸들러는 해당 예외를 우리가 정의한 `ForbiddenAccessException`으로 변환하여, 일관된 403 Forbidden 에러 응답(JSON)이 반환되도록 합니다.

jnujh committed:
fix: 메소드 시큐리티 인가 예외 처리 핸들러 추가 (#81)

[문제 원인]
@PreAuthorize 어노테이션에 의해 접근이 거부될 때, Spring Security는 `AuthorizationDeniedException`을 발생시킵니다.
이 예외를 처리하는 핸들러가 없어, 최종적으로 `500 Internal Server Error`가 응답되는 문제가 있었습니다.

[해결 방안]
GlobalExceptionHandler에 `AuthorizationDeniedException`을 처리하는 @ExceptionHandler 메소드를 추가했습니다.
이 핸들러는 해당 예외를 우리가 정의한 `ForbiddenAccessException`으로 변환하여, 일관된 403 Forbidden 에러 응답(JSON)이 반환되도록 합니다.

jnujh committed:
fix: 메소드 시큐리티 인가 예외 처리 핸들러 추가 (#81)

[문제 원인]
@PreAuthorize 어노테이션에 의해 접근이 거부될 때, Spring Security는 `AuthorizationDeniedException`을 발생시킵니다.
이 예외를 처리하는 핸들러가 없어, 최종적으로 `500 Internal Server Error`가 응답되는 문제가 있었습니다.

[해결 방안]
GlobalExceptionHandler에 `AuthorizationDeniedException`을 처리하는 @ExceptionHandler 메소드를 추가했습니다.
이 핸들러는 해당 예외를 우리가 정의한 `ForbiddenAccessException`으로 변환하여, 일관된 403 Forbidden 에러 응답(JSON)이 반환되도록 합니다.

jnujh committed:
test: 역할 기반 인가 로직에 대한 컨트롤러 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션을 사용한 역할 기반 인가 규칙이 실제 운영 환경과 동일한 조건에서 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경]
- `@WebMvcTest`의 한계(@PreAuthorize 미동작)를 해결하기 위해 `@SpringBootTest` 환경으로 전환했습니다.
- `@WithMockUser`를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
- `getClubDashboard` API에 대해 다음 시나리오를 검증합니다.
  - 성공: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
  - 실패: 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.

jnujh committed:
test: 역할 기반 인가 로직에 대한 컨트롤러 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션을 사용한 역할 기반 인가 규칙이 실제 운영 환경과 동일한 조건에서 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경]
- `@WebMvcTest`의 한계(@PreAuthorize 미동작)를 해결하기 위해 `@SpringBootTest` 환경으로 전환했습니다.
- `@WithMockUser`를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
- `getClubDashboard` API에 대해 다음 시나리오를 검증합니다.
  - 성공: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
  - 실패: 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.

jnujh committed:
test: 역할 기반 인가 로직에 대한 컨트롤러 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션을 사용한 역할 기반 인가 규칙이 실제 운영 환경과 동일한 조건에서 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경]
- `@WebMvcTest`의 한계(@PreAuthorize 미동작)를 해결하기 위해 `@SpringBootTest` 환경으로 전환했습니다.
- `@WithMockUser`를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
- `getClubDashboard` API에 대해 다음 시나리오를 검증합니다.
  - 성공: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
  - 실패: 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.

jnujh committed:
test: 역할 기반 인가 로직에 대한 컨트롤러 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션을 사용한 역할 기반 인가 규칙이 실제 운영 환경과 동일한 조건에서 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경]
- `@WebMvcTest`의 한계(@PreAuthorize 미동작)를 해결하기 위해 `@SpringBootTest` 환경으로 전환했습니다.
- `@WithMockUser`를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
- `getClubDashboard` API에 대해 다음 시나리오를 검증합니다.
  - 성공: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
  - 실패: 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.

jnujh committed:
test: ClubController에 대한 역할 기반 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션을 사용한 역할 기반 인가 규칙이 실제 운영 환경과 동일한 조건에서 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[검증 내용]
`getClubDashboard` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공 케이스: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
- 실패 케이스 (권한 부족): 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.
- 실패 케이스 (수평적 권한): 다른 동아리의 관리자가 접근 시 403 Forbidden 응답을 받는다.
- 실패 케이스 (미인증): 로그인하지 않은 사용자는 401 Unauthorized 응답을 받는다.

jnujh committed:
test: ClubController에 대한 역할 기반 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션을 사용한 역할 기반 인가 규칙이 실제 운영 환경과 동일한 조건에서 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[검증 내용]
`getClubDashboard` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공 케이스: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
- 실패 케이스 (권한 부족): 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.
- 실패 케이스 (수평적 권한): 다른 동아리의 관리자가 접근 시 403 Forbidden 응답을 받는다.
- 실패 케이스 (미인증): 로그인하지 않은 사용자는 401 Unauthorized 응답을 받는다.

jnujh committed:
test: ClubController에 대한 역할 기반 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션을 사용한 역할 기반 인가 규칙이 실제 운영 환경과 동일한 조건에서 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[검증 내용]
`getClubDashboard` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공 케이스: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
- 실패 케이스 (권한 부족): 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.
- 실패 케이스 (수평적 권한): 다른 동아리의 관리자가 접근 시 403 Forbidden 응답을 받는다.
- 실패 케이스 (미인증): 로그인하지 않은 사용자는 401 Unauthorized 응답을 받는다.

jnujh committed:
test: ApplicationController에 대한 역할 기반 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize("hasAuthority(...)") 어노테이션을 사용한 기본 역할 기반 인가 규칙이 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @WithMockUser를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
`getApplicationDetail` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
- 실패: 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.
- 실패: 다른 동아리의 관리자가 접근 시 403 Forbidden 응답을 받는다.
- 실패: 로그인하지 않은 사용자는 401 Unauthorized 응답을 받는다.

jnujh committed:
test: ApplicationController에 대한 역할 기반 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize("hasAuthority(...)") 어노테이션을 사용한 기본 역할 기반 인가 규칙이 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @WithMockUser를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
`getApplicationDetail` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
- 실패: 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.
- 실패: 다른 동아리의 관리자가 접근 시 403 Forbidden 응답을 받는다.
- 실패: 로그인하지 않은 사용자는 401 Unauthorized 응답을 받는다.

jnujh committed:
test: ApplicationController에 대한 역할 기반 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize("hasAuthority(...)") 어노테이션을 사용한 기본 역할 기반 인가 규칙이 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @WithMockUser를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
`getApplicationDetail` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공: 권한이 있는 사용자(CLUB_ADMIN)는 200 OK 응답을 받는다.
- 실패: 권한이 없는 사용자(CLUB_MEMBER)는 403 Forbidden 응답을 받는다.
- 실패: 다른 동아리의 관리자가 접근 시 403 Forbidden 응답을 받는다.
- 실패: 로그인하지 않은 사용자는 401 Unauthorized 응답을 받는다.

jnujh committed:
test: 동적 인가 테스트를 위한 @WithMockCustomUser 어노테이션 추가 (#81)

[도입 배경]
@PreAuthorize("@customSecurityService...")와 같이, PrincipalDetails의 `memberships` 맵을 직접 조회하는
동적 인가 로직을 테스트할 때, 기존 @WithMockUser는 `memberships` 정보를 제공하지 못하는 한계가 있었습니다.

[변경 내용]
1. `@WithMockCustomUser` 어노테이션을 정의하여, 테스트에서 `userId`와 `memberships` 정보를 선언적으로 설정할 수 있도록 했습니다.
2. `WithMockCustomUserSecurityContextFactory`를 구현하여, @WithMockCustomUser 어노테이션의 정보를 바탕으로
`memberships` 맵을 포함한 `PrincipalDetails` 객체를 생성하고, 이를 SecurityContext에 주입하도록 했습니다.
3. User 엔티티의 ID를 주입할 때, 엔티티의 생성 규칙을 존중하기 위해 `ReflectionTestUtils`를 사용하여 테스트의 의도를 명확히 했습니다.

[기대 효과]
이 커스텀 어노테이션을 통해, 실제 애플리케이션과 동일한 구조의 PrincipalDetails 객체를 사용하여 복잡한 동적 인가 로직을 정확하게 테스트할 수 있습니다.

jnujh committed:
test: 동적 인가 테스트를 위한 @WithMockCustomUser 어노테이션 추가 (#81)

[도입 배경]
@PreAuthorize("@customSecurityService...")와 같이, PrincipalDetails의 `memberships` 맵을 직접 조회하는
동적 인가 로직을 테스트할 때, 기존 @WithMockUser는 `memberships` 정보를 제공하지 못하는 한계가 있었습니다.

[변경 내용]
1. `@WithMockCustomUser` 어노테이션을 정의하여, 테스트에서 `userId`와 `memberships` 정보를 선언적으로 설정할 수 있도록 했습니다.
2. `WithMockCustomUserSecurityContextFactory`를 구현하여, @WithMockCustomUser 어노테이션의 정보를 바탕으로
`memberships` 맵을 포함한 `PrincipalDetails` 객체를 생성하고, 이를 SecurityContext에 주입하도록 했습니다.
3. User 엔티티의 ID를 주입할 때, 엔티티의 생성 규칙을 존중하기 위해 `ReflectionTestUtils`를 사용하여 테스트의 의도를 명확히 했습니다.

[기대 효과]
이 커스텀 어노테이션을 통해, 실제 애플리케이션과 동일한 구조의 PrincipalDetails 객체를 사용하여 복잡한 동적 인가 로직을 정확하게 테스트할 수 있습니다.

jnujh committed:
test: 동적 인가 테스트를 위한 @WithMockCustomUser 어노테이션 추가 (#81)

[도입 배경]
@PreAuthorize("@customSecurityService...")와 같이, PrincipalDetails의 `memberships` 맵을 직접 조회하는
동적 인가 로직을 테스트할 때, 기존 @WithMockUser는 `memberships` 정보를 제공하지 못하는 한계가 있었습니다.

[변경 내용]
1. `@WithMockCustomUser` 어노테이션을 정의하여, 테스트에서 `userId`와 `memberships` 정보를 선언적으로 설정할 수 있도록 했습니다.
2. `WithMockCustomUserSecurityContextFactory`를 구현하여, @WithMockCustomUser 어노테이션의 정보를 바탕으로
`memberships` 맵을 포함한 `PrincipalDetails` 객체를 생성하고, 이를 SecurityContext에 주입하도록 했습니다.
3. User 엔티티의 ID를 주입할 때, 엔티티의 생성 규칙을 존중하기 위해 `ReflectionTestUtils`를 사용하여 테스트의 의도를 명확히 했습니다.

[기대 효과]
이 커스텀 어노테이션을 통해, 실제 애플리케이션과 동일한 구조의 PrincipalDetails 객체를 사용하여 복잡한 동적 인가 로직을 정확하게 테스트할 수 있습니다.

jnujh committed:
test: 동적 인가 테스트를 위한 @WithMockCustomUser 어노테이션 추가 (#81)

[도입 배경]
@PreAuthorize("@customSecurityService...")와 같이, PrincipalDetails의 `memberships` 맵을 직접 조회하는
동적 인가 로직을 테스트할 때, 기존 @WithMockUser는 `memberships` 정보를 제공하지 못하는 한계가 있었습니다.

[변경 내용]
1. `@WithMockCustomUser` 어노테이션을 정의하여, 테스트에서 `userId`와 `memberships` 정보를 선언적으로 설정할 수 있도록 했습니다.
2. `WithMockCustomUserSecurityContextFactory`를 구현하여, @WithMockCustomUser 어노테이션의 정보를 바탕으로
`memberships` 맵을 포함한 `PrincipalDetails` 객체를 생성하고, 이를 SecurityContext에 주입하도록 했습니다.
3. User 엔티티의 ID를 주입할 때, 엔티티의 생성 규칙을 존중하기 위해 `ReflectionTestUtils`를 사용하여 테스트의 의도를 명확히 했습니다.

[기대 효과]
이 커스텀 어노테이션을 통해, 실제 애플리케이션과 동일한 구조의 PrincipalDetails 객체를 사용하여 복잡한 동적 인가 로직을 정확하게 테스트할 수 있습니다.

jnujh committed:
test: CommentController에 대한 동적 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize("@customSecurityService...")와 같이, 커스텀 빈을 이용한 동적 권한 검사 로직이 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경 및 방법]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @customSecurityService가 의존하는 `ApplicationRepository`의 DB 조회를 @MockBean으로 처리했습니다.
- 새로 추가한 `@WithMockCustomUser` 어노테이션을 사용하여, `memberships` 정보를 포함한 가짜 사용자를 생성하고 다양한 시나리오를 검증합니다.

[검증 내용]
`getComments` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공: 권한이 있는 사용자(해당 동아리 관리자)는 200 OK 응답을 받는다.
- 실패: 권한이 없는 사용자(일반 회원, 다른 동아리 관리자, 미인증 사용자)는 각각 403 또는 401 응답을 받는다.

jnujh committed:
test: CommentController에 대한 동적 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize("@customSecurityService...")와 같이, 커스텀 빈을 이용한 동적 권한 검사 로직이 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경 및 방법]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @customSecurityService가 의존하는 `ApplicationRepository`의 DB 조회를 @MockBean으로 처리했습니다.
- 새로 추가한 `@WithMockCustomUser` 어노테이션을 사용하여, `memberships` 정보를 포함한 가짜 사용자를 생성하고 다양한 시나리오를 검증합니다.

[검증 내용]
`getComments` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공: 권한이 있는 사용자(해당 동아리 관리자)는 200 OK 응답을 받는다.
- 실패: 권한이 없는 사용자(일반 회원, 다른 동아리 관리자, 미인증 사용자)는 각각 403 또는 401 응답을 받는다.

jnujh committed:
test: CommentController에 대한 동적 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize("@customSecurityService...")와 같이, 커스텀 빈을 이용한 동적 권한 검사 로직이 올바르게 동작하는지 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경 및 방법]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @customSecurityService가 의존하는 `ApplicationRepository`의 DB 조회를 @MockBean으로 처리했습니다.
- 새로 추가한 `@WithMockCustomUser` 어노테이션을 사용하여, `memberships` 정보를 포함한 가짜 사용자를 생성하고 다양한 시나리오를 검증합니다.

[검증 내용]
`getComments` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공: 권한이 있는 사용자(해당 동아리 관리자)는 200 OK 응답을 받는다.
- 실패: 권한이 없는 사용자(일반 회원, 다른 동아리 관리자, 미인증 사용자)는 각각 403 또는 401 응답을 받는다.

jnujh committed:
test: 지원서 양식(ClubApplyForm) 컨트롤러에 대한 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션이 적용된 모든 컨트롤러의 테스트 커버리지를 확보하여,
역할 기반 인가 시스템의 전체적인 안정성을 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경 및 방법]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @WithMockUser를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
`createClubApplyForm` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공 케이스: 권한이 있는 사용자(CLUB_ADMIN)는 201 Created 응답을 받는다.
- 실패 케이스: 권한이 없는 사용자(일반 회원, 다른 동아리 관리자, 미인증 사용자)는 각각 403 또는 401 응답을 받는다.

jnujh committed:
test: 지원서 양식(ClubApplyForm) 컨트롤러에 대한 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션이 적용된 모든 컨트롤러의 테스트 커버리지를 확보하여,
역할 기반 인가 시스템의 전체적인 안정성을 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경 및 방법]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @WithMockUser를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
`createClubApplyForm` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공 케이스: 권한이 있는 사용자(CLUB_ADMIN)는 201 Created 응답을 받는다.
- 실패 케이스: 권한이 없는 사용자(일반 회원, 다른 동아리 관리자, 미인증 사용자)는 각각 403 또는 401 응답을 받는다.

jnujh committed:
test: 지원서 양식(ClubApplyForm) 컨트롤러에 대한 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션이 적용된 모든 컨트롤러의 테스트 커버리지를 확보하여,
역할 기반 인가 시스템의 전체적인 안정성을 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경 및 방법]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @WithMockUser를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
`createClubApplyForm` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공 케이스: 권한이 있는 사용자(CLUB_ADMIN)는 201 Created 응답을 받는다.
- 실패 케이스: 권한이 없는 사용자(일반 회원, 다른 동아리 관리자, 미인증 사용자)는 각각 403 또는 401 응답을 받는다.

jnujh committed:
test: 지원서 양식(ClubApplyForm) 컨트롤러에 대한 인가 테스트 작성 (#81)

[테스트 목적]
@PreAuthorize 어노테이션이 적용된 모든 컨트롤러의 테스트 커버리지를 확보하여,
역할 기반 인가 시스템의 전체적인 안정성을 검증하기 위해 통합 테스트를 작성했습니다.

[테스트 환경 및 방법]
- @SpringBootTest 환경에서 실제 인가 로직이 동작하도록 구성했습니다.
- @WithMockUser를 사용하여 다양한 권한 시나리오를 시뮬레이션합니다.

[검증 내용]
`createClubApplyForm` API에 대해 다음과 같은 핵심 시나리오들을 검증합니다.
- 성공 케이스: 권한이 있는 사용자(CLUB_ADMIN)는 201 Created 응답을 받는다.
- 실패 케이스: 권한이 없는 사용자(일반 회원, 다른 동아리 관리자, 미인증 사용자)는 각각 403 또는 401 응답을 받는다.

ji-mim commented on src/main/java/com/kakaotech/team18/backend_server/global/security/JwtProvider.java:
```diff
@@ -36,14 +41,23 @@ private Key getSigningKey() {
 
         return Jwts.builder()
                 .setSubject(user.getId().toString())
+                .claim("tokenType", "ACCESS") // 토큰 타입: ACCESS
```
coderabbit이 말해준 것 같은데 "tokenType" 값이 중복해서 들어가는 것 같아서 확인 부탁드러요!

ji-mim commented on src/test/java/com/kakaotech/team18/backend_server/domain/comment/controller/CommentControllerAuthTest.java:
```diff
@@ -0,0 +1,114 @@
+        // 사용자는 1번 동아리 관리자이고, 지원서도 1번 동아리 소속이므로 200 OK를 기대합니다.
+        resultActions.andExpect(status().isOk());
+    }
+
```
운영진에 대한 테스트도 있으면 좋을 것 같아요

ji-mim reviewed (approved):
지훈님 수고많으셨어요 ! 
필수는 아닌데 혹시 여유 있으시면 테스트에 admin 말고 운영진 관련한 해피 케이스도 추가 부탁드릴게요

저는 `@PreAuthorize` 를 처음 보는데 스프링 세큐리티가 정말 많은 역할을 해주네요..! 제가 이해한게 흐름이 맞는지 확인 부탁드려요 

1. 사용자가 로그인에 성공하면 JWT 토큰을 받는데 이 값에 membership 관련된 Map 값이 포함되어 있고 Filter가 동작할 때 이 값을 기반으로 PrincipalDetails를 만들어서 Authemtication 객체에 저장해두고 이 객체를 기반으로 스프링 컨텍스트에 인가 관련된 내용이 저장된다. 
2. 이값을 토대로 clubId를 URL을 통해 알 수 있다면 @PreAuthorize의 hasAutority를 통해 인가를 확인하고 clubId를 알 수 없고 applicationId 값만 알고 있다면, isClubAdminOrExecutiveForApplication를 통해 인가를 확인한다. 
3. 테스트시에는 isClubAdminOrExecutiveForApplication를 통해 인가처리할 시 membership 값이 필요해서 @WithMockCustomerUser 어노테이션을 만들어 사용한다. 

이런 흐름이 맞을까요? 

인가 처리가  생각보다 복잡하고 시간도 많이 쓰셨을 것 같아요. 고생 많으셨어요 👍

ji-mim commented on src/main/java/com/kakaotech/team18/backend_server/global/config/SecurityConfig.java:
```diff
@@ -36,19 +37,16 @@ public SecurityFilterChain filterChain(HttpSecurity http, JwtAuthenticationFilte
+                // 지원서 양식 조회 API (공개)
+                .requestMatchers(HttpMethod.GET, "/api/clubs/*/apply").permitAll()
+                // 지원서 제출 API (공개)
+                .requestMatchers(HttpMethod.POST, "/api/clubs/*/apply-submit").permitAll()
```
이제 이 코드가 배포되면 이 API를 제외한 API에 접근할 떄는 회장이나 운영진 아이디가 필요하겠네요?

gary5876 commented on src/main/java/com/kakaotech/team18/backend_server/domain/application/controller/ApplicationController.java:
```diff
@@ -51,11 +55,15 @@ public ResponseEntity<ApplicationDetailResponseDto> getApplicationDetail(
             @ApiResponse(responseCode = "404", description = "상태를 변경할 지원서를 찾을 수 없음")
     })
-    @PatchMapping("/api/applications/{applicationId}")
+    @PreAuthorize("hasAuthority('CLUB_' + #clubId + '_CLUB_ADMIN') or hasAuthority('CLUB_' + #clubId + '_CLUB_EXECUTIVE')")
```
혹시 이것도 @customSecurityService로 하실 수 있나요?

jnujh committed:
refactor: JWT 인증 시 불필요한 DB 조회 제거로 성능 최적화

JWT 인증 필터가 매 요청마다 DB를 조회하던 비효율적인 로직을 개선합니다.
이제 토큰의 정보만으로 인증을 처리하여, 불필요한 DB 부하를 제거하고 API 응답 속도를 향상시킵니다.

•PrincipalDetails: User 엔티티 대신 userId만 갖도록 변경하여 DB 의존성을 제거.
•PrincipalDetailsService: DB 조회 로직을 삭제하고, 토큰 정보로 인증 객체를 직접 생성.
•CommentController 등: PrincipalDetails 구조 변경에 따른 연관 코드의 컴파일 에러 수정.

이로써 JWT 인증의 주된 흐름에서 DB 조회가 완전히 사라졌으며, 인증 계층과 도메인 계층의 결합도를 낮춰 코드 구조를 개선했습니다.

jnujh committed:
refactor: JWT 인증 시 불필요한 DB 조회 제거로 성능 최적화

JWT 인증 필터가 매 요청마다 DB를 조회하던 비효율적인 로직을 개선합니다.
이제 토큰의 정보만으로 인증을 처리하여, 불필요한 DB 부하를 제거하고 API 응답 속도를 향상시킵니다.

•PrincipalDetails: User 엔티티 대신 userId만 갖도록 변경하여 DB 의존성을 제거.
•PrincipalDetailsService: DB 조회 로직을 삭제하고, 토큰 정보로 인증 객체를 직접 생성.
•CommentController 등: PrincipalDetails 구조 변경에 따른 연관 코드의 컴파일 에러 수정.

이로써 JWT 인증의 주된 흐름에서 DB 조회가 완전히 사라졌으며, 인증 계층과 도메인 계층의 결합도를 낮춰 코드 구조를 개선했습니다.

jnujh committed:
refactor: JWT 인증 시 불필요한 DB 조회 제거로 성능 최적화

JWT 인증 필터가 매 요청마다 DB를 조회하던 비효율적인 로직을 개선합니다.
이제 토큰의 정보만으로 인증을 처리하여, 불필요한 DB 부하를 제거하고 API 응답 속도를 향상시킵니다.

•PrincipalDetails: User 엔티티 대신 userId만 갖도록 변경하여 DB 의존성을 제거.
•PrincipalDetailsService: DB 조회 로직을 삭제하고, 토큰 정보로 인증 객체를 직접 생성.
•CommentController 등: PrincipalDetails 구조 변경에 따른 연관 코드의 컴파일 에러 수정.

이로써 JWT 인증의 주된 흐름에서 DB 조회가 완전히 사라졌으며, 인증 계층과 도메인 계층의 결합도를 낮춰 코드 구조를 개선했습니다.

jnujh committed:
refactor: JWT 인증 시 불필요한 DB 조회 제거로 성능 최적화

JWT 인증 필터가 매 요청마다 DB를 조회하던 비효율적인 로직을 개선합니다.
이제 토큰의 정보만으로 인증을 처리하여, 불필요한 DB 부하를 제거하고 API 응답 속도를 향상시킵니다.

•PrincipalDetails: User 엔티티 대신 userId만 갖도록 변경하여 DB 의존성을 제거.
•PrincipalDetailsService: DB 조회 로직을 삭제하고, 토큰 정보로 인증 객체를 직접 생성.
•CommentController 등: PrincipalDetails 구조 변경에 따른 연관 코드의 컴파일 에러 수정.

이로써 JWT 인증의 주된 흐름에서 DB 조회가 완전히 사라졌으며, 인증 계층과 도메인 계층의 결합도를 낮춰 코드 구조를 개선했습니다.

jnujh committed:
refactor: JWT 서명 키 캐싱을 통한 생성 로직 최적화

JwtProvider가 토큰을 생성하거나 검증할 때마다 getSigningKey() 메소드 내부에서 매번 새로운 Key 객체를 생성하던 비효율을 개선합니다.

서명 키는 애플리케이션 실행 중 변경되지 않는 불변(Immutable) 값에 가까우므로,
@PostConstruct를 사용하여 JwtProvider Bean이 초기화될 때 서명 키를 한 번만 생성하고 필드에 캐싱하도록 변경했습니다.

이를 통해 불필요한 객체 생성을 방지하고, 모든 토큰 관련 작업에서 미세한 성능 향상을 기대할 수 있습니다.

jnujh committed:
refactor: JWT 서명 키 캐싱을 통한 생성 로직 최적화

JwtProvider가 토큰을 생성하거나 검증할 때마다 getSigningKey() 메소드 내부에서 매번 새로운 Key 객체를 생성하던 비효율을 개선합니다.

서명 키는 애플리케이션 실행 중 변경되지 않는 불변(Immutable) 값에 가까우므로,
@PostConstruct를 사용하여 JwtProvider Bean이 초기화될 때 서명 키를 한 번만 생성하고 필드에 캐싱하도록 변경했습니다.

이를 통해 불필요한 객체 생성을 방지하고, 모든 토큰 관련 작업에서 미세한 성능 향상을 기대할 수 있습니다.

jnujh committed:
refactor: JWT 서명 키 캐싱을 통한 생성 로직 최적화

JwtProvider가 토큰을 생성하거나 검증할 때마다 getSigningKey() 메소드 내부에서 매번 새로운 Key 객체를 생성하던 비효율을 개선합니다.

서명 키는 애플리케이션 실행 중 변경되지 않는 불변(Immutable) 값에 가까우므로,
@PostConstruct를 사용하여 JwtProvider Bean이 초기화될 때 서명 키를 한 번만 생성하고 필드에 캐싱하도록 변경했습니다.

이를 통해 불필요한 객체 생성을 방지하고, 모든 토큰 관련 작업에서 미세한 성능 향상을 기대할 수 있습니다.

ji-mim committed:
feature: commandLineRunner 적용 (#151)

ji-mim committed:
refactor: commandLineRunner 설정 프로필 변경 및 data.sql 미사용 (#176)

- data.sql이 적용되지 않도록 이름을 변경했습니다.
- 또한 CommandLineRunner가 적용될 프로필을 디폹트(로컬)과 prod(운영)으로 설정했습니다.
- CommandLineRunner가 실행될 때 insert 로그가 엄청 찍히는데 추후에 운영할 때는 쿼리가 콘솔에 안찍히게 변경해야할 것 같습니다. 일단 docker 컴포즈 파일에 로그 관련 설정을 추가해 로그가 용량을 많이 차지하는걸 방지해두었습니다.

ji-mim committed:
remove: Notification 엔티티 제거 (#176)

ji-mim committed:
feat: favicon.ico 추가 (#176)

- favicon 에러가 계속 떠서 간단하게 하나 만들어 추가했습니다.

ji-mim committed:
refactor: 초기데이터 경량화 (#176)

jnujh committed:
Merge branch 'develop' into feature/role-based-authorization#81

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/domain/application/repository/ApplicationRepository.java
#	src/main/java/com/kakaotech/team18/backend_server/domain/clubMember/repository/ClubMemberRepository.java
#	src/main/java/com/kakaotech/team18/backend_server/global/config/SecurityConfig.java
#	src/main/java/com/kakaotech/team18/backend_server/global/exception/handler/GlobalExceptionHandler.java
#	src/main/java/com/kakaotech/team18/backend_server/global/security/JwtAuthenticationFilter.java

jnujh committed:
fix: ClubApplyFormRequestDto 생성자 인자 추가

jnujh committed:
fix: ClubApplyFormRequestDto 생성자 인자 추가

jnujh committed:
refactor: 댓글 서비스 테스트를 단위/통합 테스트로 분리 (#81)

기존 CommentServiceImplTest가 단위 테스트와 통합 테스트의 역할을 겸하면서 발생하던 문제를 해결

문제점
•@SpringBootTest로 작성된 CommentServiceImplTest가 다른 통합 테스트(ClubRestClientTest 등)와 컨텍스트 설정이 달라
'All Tests' 실행 시 컨텍스트 재시작 및 로그 오염으로 인한 테스트 실패를 유발했습니다.

•순수 로직을 검증해야 할 단위 테스트에 Spring AOP 기능(@Retryable) 테스트와 로그 출력 검증 로직이 포함되어
테스트가 무겁고 깨지기 쉬운(brittle) 문제가 있었습니다.

변경 내용
1.CommentServiceImplTest를 순수 단위 테스트로 전환
•@SpringBootTest를 제거하고 @ExtendWith(MockitoExtension.class)를 사용하여 Spring 컨텍스트 의존성을 완전히 제거했습니다.
•로그 출력 검증(CapturedOutput)을 제거하고, verify를 사용한 행위 검증에 집중하도록 수정했습니다.

2.CommentServiceIntegrationTest 신규 생성
•@Retryable과 같이 Spring 프레임워크 기능에 의존하는 테스트를 위해 새로운 서비스 계층 통합 테스트 파일을 생성했습니다.
•기존 CommentServiceImplTest에 있던 @Retryable 테스트 케이스를 이곳으로 이전하여 테스트 커버리지 공백을 방지했습니다.

jnujh committed:
refactor: 댓글 서비스 테스트를 단위/통합 테스트로 분리 (#81)

기존 CommentServiceImplTest가 단위 테스트와 통합 테스트의 역할을 겸하면서 발생하던 문제를 해결

문제점
•@SpringBootTest로 작성된 CommentServiceImplTest가 다른 통합 테스트(ClubRestClientTest 등)와 컨텍스트 설정이 달라
'All Tests' 실행 시 컨텍스트 재시작 및 로그 오염으로 인한 테스트 실패를 유발했습니다.

•순수 로직을 검증해야 할 단위 테스트에 Spring AOP 기능(@Retryable) 테스트와 로그 출력 검증 로직이 포함되어
테스트가 무겁고 깨지기 쉬운(brittle) 문제가 있었습니다.

변경 내용
1.CommentServiceImplTest를 순수 단위 테스트로 전환
•@SpringBootTest를 제거하고 @ExtendWith(MockitoExtension.class)를 사용하여 Spring 컨텍스트 의존성을 완전히 제거했습니다.
•로그 출력 검증(CapturedOutput)을 제거하고, verify를 사용한 행위 검증에 집중하도록 수정했습니다.

2.CommentServiceIntegrationTest 신규 생성
•@Retryable과 같이 Spring 프레임워크 기능에 의존하는 테스트를 위해 새로운 서비스 계층 통합 테스트 파일을 생성했습니다.
•기존 CommentServiceImplTest에 있던 @Retryable 테스트 케이스를 이곳으로 이전하여 테스트 커버리지 공백을 방지했습니다.

departmentowl committed:
Merge remote-tracking branch 'origin' into refactor/apply-commandlinerunner#176

departmentowl committed:
chore:edited data.sql(#176)

jnujh committed:
fix: ClubRestClientTest에 테스트 전용 보안 설정을 적용하여 실패 해결 (#81)

역할 기반 인가 시스템이 도입된 후, 보안이 적용된 API에 인증 없이 접근하여 401 에러로 실패하던 ClubRestClientTest를 수정했습니다.

문제점
•ClubRestClientTest의 목적은 파일 업로드 크기 제한 정책을 검증하는 것이지만, 인가 시스템 도입 후 인증 문제로 인해 테스트가 실패했습니다.
•외부 TestSecurityConfig를 @Import하는 방식은 @SpringBootTest 환경에서 메인 SecurityConfig와 충돌하거나,
우선순위 문제로 인해 제대로 적용되지 않는 문제를 야기했습니다.

해결 방안
•@Import 방식을 제거하고, ClubRestClientTest 클래스 내부에 static 중첩 클래스로 테스트 전용 TestSecurityConfig를 정의했습니다.
•이 방식은 Spring Boot 테스트에서 특정 Bean(여기서는 SecurityFilterChain)을 명시적으로 덮어쓰는 가장 확실하고 표준적인 방법입니다.
•spring.main.allow-bean-definition-overriding=true 속성을 유지하여,
테스트용 보안 설정이 애플리케이션의 기본 보안 설정을 안전하게 대체하도록 허용했습니다.

기대 효과
•ClubRestClientTest가 더 이상 인가(Authorization)의 영향을 받지 않고, 본래 목적인 파일 업로드 정책 검증에만 집중할 수 있게 되었습니다.

jnujh committed:
fix: ClubRestClientTest에 테스트 전용 보안 설정을 적용하여 실패 해결 (#81)

역할 기반 인가 시스템이 도입된 후, 보안이 적용된 API에 인증 없이 접근하여 401 에러로 실패하던 ClubRestClientTest를 수정했습니다.

문제점
•ClubRestClientTest의 목적은 파일 업로드 크기 제한 정책을 검증하는 것이지만, 인가 시스템 도입 후 인증 문제로 인해 테스트가 실패했습니다.
•외부 TestSecurityConfig를 @Import하는 방식은 @SpringBootTest 환경에서 메인 SecurityConfig와 충돌하거나,
우선순위 문제로 인해 제대로 적용되지 않는 문제를 야기했습니다.

해결 방안
•@Import 방식을 제거하고, ClubRestClientTest 클래스 내부에 static 중첩 클래스로 테스트 전용 TestSecurityConfig를 정의했습니다.
•이 방식은 Spring Boot 테스트에서 특정 Bean(여기서는 SecurityFilterChain)을 명시적으로 덮어쓰는 가장 확실하고 표준적인 방법입니다.
•spring.main.allow-bean-definition-overriding=true 속성을 유지하여,
테스트용 보안 설정이 애플리케이션의 기본 보안 설정을 안전하게 대체하도록 허용했습니다.

기대 효과
•ClubRestClientTest가 더 이상 인가(Authorization)의 영향을 받지 않고, 본래 목적인 파일 업로드 정책 검증에만 집중할 수 있게 되었습니다.

jnujh committed:
Merge branch 'feature/role-based-authorization#81' of https://github.com/kakao-tech-campus-3rd-step3/Team18_BE into feature/role-based-authorization#81

# Conflicts:
#	src/test/java/com/kakaotech/team18/backend_server/domain/clubApplyForm/controller/ClubApplyFormControllerAuthTest.java

departmentowl committed:
feat:datailitializer create(#176)

departmentowl committed:
chore : deleted notification from data.sql(#176)

departmentowl committed:
chore: fixed typo from club_member in data.sql(#176)

jnujh committed:
fix: 컨트롤러 경로 버그 수정 및 인가 기능 추가 (#81)

PR 리뷰 피드백을 반영하여 ApplicationController에 존재하던 버그와 보안 취약점을 해결합니다.

1.API 경로 중복 문제 해결
•클래스 레벨의 @RequestMapping("/api/clubs")와 메서드 레벨의 @PatchMapping 경로가 중복되어,
API 주소가 /api/clubs/api/clubs/...로 잘못 생성되던 버그를 수정했습니다.
•메서드 매핑에서 중복 경로를 제거하여 올바른 API 주소로 매핑되도록 변경했습니다.

2.합/불 처리 API 인가 누락 문제 해결
•지원자 합/불 처리를 하는 sendPassFailMessage 메서드에 @PreAuthorize가 누락되어 있던 보안 취약점을 수정했습니다.
•#clubId를 동적으로 참조하여, 해당 동아리의 CLUB_ADMIN 또는 CLUB_EXECUTIVE 역할(Authority)을 가진 사용자만 해당 API를 호출할 수 있도록 인가 규칙을 추가했습니다.

jnujh committed:
fix: 오타 수정 (#81)

gary5876 committed:
Merge pull request #187 from kakao-tech-campus-3rd-step3/refactor/apply-commandlinerunner#176

[REFACTOR] data.sql 에서 CommandLineRunner로 데이터 삽입 변경 (#176)

jnujh committed:
fix: 중복 클레임 제거 (#81)

tokenType 두 번 설정됨

jnujh committed:
test: CLUB_EXECUTIVE 커버 테스트 추가 (#81)

ji-mim committed:
refactor: data.sql 활성화

ji-mim committed:
[MERGE] data.sql 활성화

[FIX] data.sql 활성화

jnujh committed:
refactor: PreAuthorize SpEL을 CustomSecurityService 메서드로 리팩토링 (#81)

ApplicationController의 인가 로직을 개선하여 가독성, 재사용성, 유지보수성을 향상시켰습니다.

변경 내용
1.CustomSecurityService에 신규 메서드 추가
•clubId를 파라미터로 받아, DB 조회 없이 JWT 클레임 정보만으로
동아리 관리자(회장 또는 운영진) 권한을 검사하는 isClubAdminOrExecutive(Long clubId) 메서드를 추가했습니다.
•메서드의 역할과 동작을 명확히 하기 위해 상세한 Javadoc과 인라인 주석을 추가했습니다.

2.ApplicationController 리팩토링
•updateApplicationStatus 메서드의 @PreAuthorize에 사용되던 길고 복잡한 SpEL 표현식을 제거했습니다.
•새로 만든 @customSecurityService.isClubAdminOrExecutive(#clubId)를 호출하는 방식으로 변경

jnujh committed:
fix: 존재하지 않는 사용자에 대한 인증 통과 보안 취약점 해결 (#81)

PrincipalDetailsService에서 JWT의 userId를 검증 없이 신뢰하여, 탈퇴했거나 존재하지 않는 사용자로도 인증이 통과되는 심각한 보안 취약점을 해결합니다.

변경 내용
1.PrincipalDetailsService 보안 강화
•loadUserByUsername 메서드 내에서 PrincipalDetails를 생성하기 전에,
userRepository.existsById(userId)를 호출하여 데이터베이스에 해당 사용자가 실제로 존재하는지 확인하는 검증 로직을 추가했습니다.
•존재하지 않는 userId일 경우, UsernameNotFoundException을 발생시켜 인증을 즉시 실패 처리합니다.

2.PrincipalDetailsServiceTest 단위 테스트 추가
•수정된 로직을 검증하기 위해 순수 단위 테스트 클래스를 신규로 생성했습니다.
•userRepository.existsById()가 false를 반환하는 경우, UsernameNotFoundException이 정상적으로 발생하는지
확인하는 테스트 케이스를 추가하여 보안 강화 조치가 올바르게 동작함을 보장합니다.

기대 효과
•탈퇴한 사용자의 만료되지 않은 토큰을 이용한 인증 시도를 차단합니다.
•존재하지 않는 userId를 포함한 토큰으로 인증을 시도하는 비정상적인 접근을 방지하여 시스템 보안을 강화합니다.

jnujh committed:
fix: 주석 수정 (#81)

jnujh committed:
fix: 메소드명 수정 (#81)

jnujh committed:
fix: 비활성 멤버에게 권한이 부여되는 보안 결함 수정 (#81)

JWT 토큰 생성 시, 비활성(INACTIVE) 상태의 동아리 멤버십까지 포함하여 부적절한 권한이 부여되던 심각한 보안 결함을 해결합니다.

변경 내용
1.ClubMemberRepository 쿼리 수정
•findClubMembershipsByUser 메서드의 JPQL 쿼리에 activeStatus = ACTIVE 조건을 추가했습니다.
•이를 통해 JWT 클레임에는 오직 현재 활동 중인 멤버십 정보만 포함되도록 수정했습니다.
•불필요한 @Param 어노테이션을 제거하여 코드 일관성을 개선했습니다.

2.ClubMemberRepositoryTest 통합 테스트 추가
•수정된 쿼리의 동작을 검증하기 위해 @DataJpaTest 기반의 Repository 테스트 클래스를 신규로 생성했습니다.
•ACTIVE 상태의 멤버십은 조회되고, INACTIVE 상태의 멤버십은 필터링되는지를 명시적으로 검증하는 테스트 케이스를 추가하여 로직의 정확성을 보장합니다.

기대 효과
•동아리 활동을 중단한 비활성 회원이 더 이상 관리자 권한 등을 획득할 수 없도록 하여, 시스템의 인가(Authorization) 보안을 강화합니다.

jnujh committed:
fix: 중복 제거

jnujh committed:
refactor: 로그 표시되도록 수정

jnujh committed:
Merge branch 'feature/role-based-authorization#81' of https://github.com/kakao-tech-campus-3rd-step3/Team18_BE into feature/role-based-authorization#81

# Conflicts:
#	src/main/java/com/kakaotech/team18/backend_server/domain/application/controller/ApplicationController.java
#	src/main/java/com/kakaotech/team18/backend_server/domain/clubMember/repository/ClubMemberRepository.java
#	src/main/java/com/kakaotech/team18/backend_server/global/config/SecurityConfig.java
#	src/main/java/com/kakaotech/team18/backend_server/global/security/CustomSecurityService.java
#	src/main/java/com/kakaotech/team18/backend_server/global/security/JwtProvider.java
#	src/main/java/com/kakaotech/team18/backend_server/global/security/PrincipalDetailsService.java
#	src/test/java/com/kakaotech/team18/backend_server/domain/club/controller/ClubControllerAuthTest.java
#	src/test/java/com/kakaotech/team18/backend_server/domain/comment/controller/CommentControllerAuthTest.java
#	src/test/java/com/kakaotech/team18/backend_server/domain/comment/service/CommentServiceImplTest.java
#	src/test/java/com/kakaotech/team18/backend_server/domain/comment/service/CommentServiceIntegrationTest.java

jnujh committed:
fix: 오타 수정 (#81)

jnujh committed:
fix: 지원서 상태 변경 API의 수평적 권한 상승 취약점 해결 (#81)

updateApplicationStatus API에서 발생할 수 있는 심각한 수평적 권한 상승(Horizontal Privilege Escalation) 보안 취약점을 해결합니다.

문제점
•기존 @PreAuthorize 인가 로직이 URL 경로의 clubId만을 기준으로 권한을 검사했습니다.
•이로 인해, 특정 동아리의 관리자가 URL의 clubId는 자신의 것으로, applicationId는 다른 동아리의 것으로 조합하여 요청을 보내면,
다른 동아리의 지원서 상태를 수정할 수 있는 치명적인 취약점이 있었습니다.

해결 방안
•@PreAuthorize의 검사 기준을 clubId에서 실제 수정 대상인 applicationId로 변경했습니다.
•이제 @customSecurityService.isClubAdminOrExecutiveForApplication(#applicationId)를 통해,
현재 사용자가 해당 지원서가 속한 동아리의 관리자인지를 직접 검증합니다.
•서비스 계층의 변경 없이, 컨트롤러의 인가 계층에서만 취약점을 해결하는 방향으로 수정했습니다.

기대 효과
•다른 동아리의 지원서 상태를 임의로 수정할 수 있었던 수평적 권한 상승 공격 경로를 원천적으로 차단하여 시스템 보안을 강화합니다.

jnujh committed:
fix: CustomSecuritySevice 사용하도록 수정 (#81)

jnujh committed:
fix: CustomSecurityService 사용하도록 수정한 것에 맞게 테스트 수정 (#81)

jnujh committed:
fix: 에러 해결을 위한 인가 적용 (#81)
